### PR TITLE
Instrumentation: Removes invalid chars from label names

### PR DESCRIPTION
**Pull request by bergquist**
skip instrumenting datasources if datasource
names only contains invalid chars

### CODE REVIEW

File: pkg/infra/metrics/metrics.go [hunk 1]
```diff
@@ -589,3 +591,25 @@ func newCounterStartingAtZero(opts prometheus.CounterOpts, labelValues ...string
...
+
+	out := strings.Builder{}
+	for i, b := range name {
+		if (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || b == '_' || (b >= '0' && b <= '9' && i > 0) {
+			out.WriteRune(b)
+		}
+	}
+
+	if out.Len() == 0 {
+		return "", errors.New("label name only contains invalid chars")
```
Comment (aknuds1) [inline pkg/infra/metrics/metrics.go:611]: IMO, including the erroneous name would help in debugging?
```suggestion
		return "", fmt.Errorf("label name only contains invalid chars: %q", name)
```

File: pkg/infra/metrics/metrics_test.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+		{input: "", expected: "", err: true},
+		{input: ";;;", expected: "", err: true},
+	}
+
+	for _, tc := range testcases {
+		got, err := SanitizeLabelName(tc.input)
+		if tc.err {
+			assert.Error(t, err)
+		} else {
+			assert.Equal(t, tc.expected, got)
```
Comment (aknuds1) [inline pkg/infra/metrics/metrics_test.go:26]: I would also make sure that `err` didn't happen:
```suggestion
			require.NoError(t, err)
			assert.Equal(t, tc.expected, got)
```

**Review by aknuds1 — APPROVED** [review]
LGTM, but see comments.

**Comment by aknuds1** [comment]
Should we add the cherry-pick needed label? Also didn't know we were making more 7.1.x releases.

File: pkg/infra/metrics/metrics_test.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+)
+
+func TestLabelNameSanitization(t *testing.T) {
+	testcases := []struct {
+		input    string
+		expected string
+		err      bool
+	}{
+		{input: "job", expected: "job"},
+		{input: "job._loal['", expected: "job_loal"},
```
Comment (marefr) [inline pkg/infra/metrics/metrics_test.go:16]: Does it work with spaces as well? Personally would replace spaces with underscore (_). WDYT?

```suggestion
		{input: "job._loal['", expected: "job_loal"},
		{input: "Data source'", expected: "Datasource"},
```

File: pkg/models/datasource_cache.go [hunk 4]
```diff
@@ -72,7 +73,14 @@ type dataSourceTransport struct {
 
 func instrumentRoundtrip(datasourceName string, next http.RoundTripper) promhttp.RoundTripperFunc {
 	return promhttp.RoundTripperFunc(func(r *http.Request) (*http.Response, error) {
-		datasourceLabel := prometheus.Labels{"datasource": datasourceName}
+		datasourceLabelName, err := metrics.SanitizeLabelName(datasourceName)
```
Comment (marefr) [inline pkg/models/datasource_cache.go:76]: Maybe we should lowercase this label?

**Review by marefr — CHANGES_REQUESTED** [review]
LGTM, but added a couple comments/suggestions

File: pkg/models/datasource_cache.go (hunk 4 above)
Comment (bergquist) [reply pkg/models/datasource_cache.go:76]: Uppercase is allowed in Prometheus https://github.com/prometheus/common/blob/master/model/labels.go#L86

Lowercasing would increase the chance of collision so I would prefer to not do that.

File: pkg/infra/metrics/metrics_test.go (hunk 3 above)
Comment (bergquist) [reply pkg/infra/metrics/metrics_test.go:16]: Agree.

**Review by marefr — APPROVED** [review]
LGTM

**Pull request merged**